handle negative values for shapes in PMMatrix

rakki-18 opened the pull request:
This PR solves the issue - #205 partially.
Changes introduced with this PR :
- When negative values are given as input to the shapes of a PMMatrix, an AssertionFailure error is returned.

hemalvarambhia commented on src/Math-Tests-Matrix/PMMatrixTest.class.st:
```diff
@@ -475,7 +475,10 @@ PMMatrixTest >> testMatrixInitialize [
-	self assert: a numberOfColumns equals: 2
+	self assert: a numberOfColumns equals: 2.
+	
+	self should: [ PMMatrix new: -2 ] raise: AssertionFailure .
```
These, according to the project guidelines, should be a separate test method e.g. `testMatrixInitialisationWithNegativeDimensions` as this is an edge case.

hemalvarambhia commented on src/Math-Matrix/PMMatrix.class.st:
```diff
@@ -520,7 +521,8 @@ PMMatrix >> initializeRows: anArrayOrVector [
 	"Build empty components for a matrix."
-		
+	[ rowsInteger  isInteger and: [ rowsInteger   >= 0 ] ] assert.
+	[ columnsInteger  isInteger and: [ columnsInteger  >= 0 ] ] assert.
```
What failure message do clients get here? I'd it easy to debug from?

rakki-18 replied:
It shows up as an assertion failure only.

hemalvarambhia replied:
Would you mind posting it here? It might be worth documenting that message in the test as an alternative.

rakki-18 replied:
`AssertionFailure: Assertion failed`

hemalvarambhia replied:
Perhaps, then, we should be clearer e.g. 'Dimension of a matrix must be greater than zero'

hemalvarambhia commented on src/Math-Matrix/PMMatrix.class.st:
```diff
@@ -508,6 +508,7 @@ PMMatrix >> hash [
 { #category : #initialization }
 PMMatrix >> initialize: anInteger [
 	"Build empty components for a square matrix. No check is made: components are assumed to be orgainized in rows."
+	[ anInteger isInteger and: [ anInteger  >= 0 ] ] assert.
```
Can we have a zero dimensional matrix? I don't think we do.

rakki-18 replied:
I am not sure of this actually. But `Array2D` allows zero-dimensional matrix and we can create a `PMVector` of size 0. So, I felt we should allow zero-dimensional matrix in `PMMatrix`. What do you think?

hemalvarambhia replied:
This is a domain question that may be worth raising with Serge or in discord

hemalvarambhia replied:
Maybe the initialize method should be renamed after the discord conversation, and the comment for this method, which says clients get a square matrix

SergeStinckwich commented:
Rename PMatrix>>initialize: as PMMatrix>>initializeSquare:
Use PMatrix>>rows:columns for PMMatrix>>initializeSquare: to avoid code duplication
Add tests to show these methods are working

hemalvarambhia commented on src/Math-Matrix/PMMatrix.class.st:
```diff
@@ -105,7 +105,7 @@ PMMatrix class >> lupCRLCriticalDimension [
 	"Create an empty square matrix of dimension anInteger."
 
-	^ self new initialize: anInteger
+	^ self new initializeSquare: anInteger
```
`dimension` rather than `anInteger`

hemalvarambhia commented on src/Math-Matrix/PMMatrix.class.st:
```diff
@@ -520,10 +514,17 @@ PMMatrix >> initializeRows: anArrayOrVector [
 ]
 
+{ #category : #initialization }
+PMMatrix >> initializeSquare: anInteger [
```
dimension rather than anInteger

rakki-18 commented:
@hemalvarambhia @SergeStinckwich, I have made the changes. Can you please check now?

hemalvarambhia commented on src/Math-Matrix/PMMatrix.class.st:
```diff
@@ -102,10 +102,10 @@ PMMatrix class >> lupCRLCriticalDimension [
 
 { #category : #'instance creation' }
-PMMatrix class >> new: anInteger [
+PMMatrix class >> new: dimension [
```
Correct the variable name in the comment, if you don't mind.

rakki-18 replied:
Yeah, done.

hemalvarambhia reviewed (dismissed):
Thank you for trying out our suggestions. Code looks a lot better.

hemalvarambhia commented:
@rakki-18 My colleague @SergeStinckwich suggested you might write programmer tests for the new messages & methods you introduced. I realised you are still working on that, so have dismissed my review. When those tests have been written, let us know on Discord.

rakki-18 commented:
@hemalvarambhia @SergeStinckwich The changes that I introduced:
- refactored and renamed `inititialize` method to `initializeSquare`. Test for this - [testMatrixInitializeSquare](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR482)
- added a new assertion statement that prevents initialization of PMMatrix with non-positive dimensions. Test for this - [testMatrixInitializeWithNegativeDimensions](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR474)
- Improved variable name in the PMMatrix `new:` method. Test for this - [testMatrixNew](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR541) (Although, I am not sure if this test is required as it is a bit redundant to `testMatrixInitializeSquare`)
